### PR TITLE
Allow empty query in  workspace/symbol

### PR DESCRIPTION
**Pull request by ayoub-benali**
Since LSP 3.15 workspace/symbol should allow sending empty query

https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#workspace_symbol

fix https://github.com/sublimelsp/LSP/issues/1008

### CODE REVIEW

File: plugin/workspace_symbol.py [hunk 1]
```diff
@@ -10,9 +10,6 @@
 
 class SymbolQueryInput(sublime_plugin.TextInputHandler):
 
-    def validate(self, txt: str) -> bool:
-        return txt != ""
-
     def placeholder(self) -> str:
         return "Symbol"
```
Comment (rchl) [inline plugin/workspace_symbol.py:14]: Maybe placeholder could indicate that empty string is allowed. Something like `Symbol (leave empty to request all symbols)`

**Comment by rwols** [comment]
It's weird that there's no additional server capability for this. How can we detect servers that support v3.15 of the protocol versus servers that support < 3.15?

**Comment by rwols** [comment]
Can we take this opportunity to start a discussion about workspace/symbol?

Perhaps we can do the request with an empty query, always. This way we utilize ST’s fuzzy string matching better, and it would feel more like “Goto Symbol In Project” (ctrl+shift+R)

Meaning: just like we “take over” Goto definition with F12, we may “take over” Goto Symbol In Project with ctrl+shift+R and it should act/present approximately the same.

**Comment by rchl** [comment]
> Perhaps we can do the request with an empty query, always.

Just quick thoughts without having much experience with that functionality:
1) Would we be able to just directly to symbol, without showing the list, if there is only one matching symbol?
2) I understand that `workspace/symbol` is not supported by that many servers. So would it be feasible to use it instead of `textDocument/definition`? Or if not, would it be possible to fallback to it if `workspace/symbol` doesn't work?

**Comment by ayoub-benali** [comment]
> Perhaps we can do the request with an empty query, always.

I don't know about the other servers but for Metals the search is done across the "classpath": project code, the dependencies source code, the dependencies of the dependencies etc. So you can see that sending an empty query can produce a big latency, a quite huge result list and I have no clue how servers and sublime would react to all project symbols.

> Would we be able to just directly jump to symbol, without showing the list, if there is only one matching symbol?

Good idea

> I understand that workspace/symbol is not supported by that many servers. So would it be feasible to use it instead of textDocument/definition? Or if not, would it be possible to fallback to it if workspace/symbol doesn't work?

`textDocument/definition` and `workspace/symbol` don't have the same cost and latency for servers and they don't mean the same thing semantically so I would keep distinct.

**Comment by ayoub-benali** [comment]
> It's weird that there's no additional server capability for this. 

The 3.15 spec mentions `WorkspaceSymbolClientCapabilities`  but for me it looks the same as < 3.15.

> How can we detect servers that support v3.15 of the protocol versus servers that support < 3.15?

The only difference seems to be allowing empty query so no need to detect 3.15 servers, or an I missing something ?

**Comment by rwols** [comment]
It might be possible that servers that support < 3.15 will assume the query-string is non-empty (and might crash if provided with an empty query string). But I might be splitting hairs here.

> textDocument/definition and workspace/symbol don't have the same cost and latency for servers and they don't mean the same thing semantically so I would keep distinct.

Indeed. `textDocument/definition` is a local (=easy) problem, `workspace/symbol` is a global (=hard) problem.

File: plugin/workspace_symbol.py [hunk 2]
```diff
@@ -27,24 +24,30 @@ def _format(self, s: Dict[str, Any]) -> str:
...
         if response:
             matches = response
-            choices = list(map(lambda s: self._format(s), matches))
-            window = self.view.window()
-            if window:
-                window.show_quick_panel(choices, lambda i: self._open_file(matches, i))
+            if len(matches) == 1:
+                self._open_file(matches[0])
+            else:
+                choices = list(map(lambda s: self._format(s), matches))
```
Comment (rwols) [inline plugin/workspace_symbol.py:47]: You can omit the lambda: `list(map(self._format, matches))`

**Comment by rwols** [comment]
Now that you're busy in this file anyway, perhaps you can return a `List[str]` from `_format` instead of a `str`. The `show_quick_panel` method can also take a `List[List[str]]` (instead of a `List[str]` as we pass it now). This makes the quick panel show items with three rows. The lower two rows have a small font. I think you can stash all of the semantic info there. That might be more informative. Because right now I don't quite understand what some of the results mean when I use this feature.

I skimmed through #501 but I did not find a reason for omitting this feature, perhaps we just forgot about it?

**Comment by rwols** [comment]
I've tested this locally and it doesn't actually run with an empty query string :) That's because there's still an `if` check in the `run` method.

**Comment by ayoub-benali** [comment]
Good catch thanks

**Comment by rwols** [comment]
I'm doubting whether what we do here makes any sense at all. For instance, clangd may sends hundreds of megabytes of JSON data when doing an empty query request.

What probably needs to happen is to use a `partialResultToken` to enable partial results (and a `workDoneProgressToken` while we're at it). Then, the server starts streaming results to us, and the user starts typing at the same time. It would remove some of the enormous wait time.

This is a complicated request.

**Comment by rwols** [comment]
I've been chatting in the clangd Discord and they're telling me that partial results would still kind of suck, because the editor cannot provide feedback on what the user is typing. So a server might stream useless chunks of data (e.g. the user has typed `foo::` while the server streams thousands of tokens that start with `bar::`). But it would be a step in the right direction.

**Comment by ayoub-benali** [comment]
> I'm doubting whether what we do here makes any sense at all. For instance, clangd may send hundreds of megabytes of JSON data when doing an empty query request.

This what I meant earlier about sending the command always with empty query.
For my usage it doesn't make send `workspace/symbol` with empty param but this PR is about allowing following the LSP spec. 

If you think it doesn't make sense to have such a feature let's close this PR and the related issue.

**Comment by rwols** [comment]
Let's revisit this problem when we have `workDoneProgress` and `partialResult` for requests.

**Pull request closed without merging**